src/credential/credentialController.ts

```diff
@@ -4,24 +4,27 @@ import * as template from 'url-template';
 import { UserInfo, DocsSignInStatus, EXTENSION_ID, uriHandler } from '../shared';
 import extensionConfig from '../config';
 import { parseQuery, delay, trimEndSlash } from '../utils/utils';
-import { UserSigningIn, UserSignInSucceeded, UserSignedOut, CredentialReset, UserSignInFailed, BaseEvent, UserSignInProgress, CredentialRetrieveFromLocalCredentialManager } from '../common/loggingEvents';
+import { UserSignInSucceeded, UserSignedOut, CredentialReset, UserSignInFailed, BaseEvent, UserSignInProgress, CredentialRetrieveFromLocalCredentialManager, UserSignInTriggered } from '../common/loggingEvents';
 import { EventType } from '../common/eventType';
 import { EventStream } from '../common/eventStream';
 import { KeyChain } from './keyChain';
 import { EnvironmentController } from '../common/environmentController';
+import { TimeOutError } from '../error/timeOutError';
+import { DocsError } from '../error/docsError';
+import { ErrorCode } from '../error/errorCode';
 
 async function handleAuthCallback(callback: (uri: vscode.Uri, resolve: (result: any) => void, reject: (reason: any) => void) => void): Promise<any> {
     let uriEventListener: vscode.Disposable;
     return Promise.race([
-        delay(extensionConfig.SignInTimeOut, new Error(`Timeout`)),
+        delay(extensionConfig.SignInTimeOut, new TimeOutError('Time out')),
         new Promise((resolve: (result: any) => void, reject: (reason: any) => void) => {
             uriEventListener = uriHandler.event((uri) => callback(uri, resolve, reject));
         }).then(result => {
             uriEventListener.dispose();
             return result;
         }).catch(err => {
             uriEventListener.dispose();
-            throw err;
+            return err;

```

Comment:
why not throw here and catch in the caller? Then the return can just be `result`

Reply:
just want to keep consistent with the `delay` function, which returns `TimeOut` error instead of throw, but if you think `throw` is a best practise, I can change both of them to throw.

Reply:
I would prefer `throw/catch`, this could work only because it is dynamic language

Reply:
Make sence